CHANGES.md

```diff
@@ -1,5 +1,124 @@
 # Changes
 
+## stable-2.12.0
+
+This release introduces route-based policy to Linkerd, allowing users to define
+and enforce authorization policies based on HTTP routes in a fully zero-trust
+way. These policies are built on Linkerd's strong workload identities, secured
+by mutual TLS, and configured using types from Kubernetes's [Gateway API](https://gateway-api.sigs.k8s.io/).
+
+The 2.12 release also introduces access logging, a long-awaited feature that
+optionally allows Linkerd to produce Apache-style request logs, optional
+support for `iptables-nft`, and a host of other improvements and performance
+enhancements.
+
+**Upgrade notes**: Please see the [upgrade instructions](upgrade-2120). <!-- TODO fix link -->
+
+* Proxy
+  * Added a `config.linkerd.io/shutdown-grace-period` annotation to configure
+   the proxy's maximum grace period for graceful shutdown
+  * Added a new `iptables-nft` mode for `proxy-init`
+  * Fixed an issue where the proxy-injector would break when using
+    `nodeAffinity` values for the control plane
+
+* Control Plane
+  * Added support for per-route policy by supporting AuthorizationPolicy
+    resources which target HttpRoute resources
+  * Removed kube-system exclusions from watchers to fix service discovery
+    for workloads in the kube-system namespace (thanks @JacobHenner)
+  * Updated healthcheck to ignore `Terminated` state for pods (thanks
+    @AgrimPrasad!)
```

Comment:
nit: for the contributor thank you we have an exclamation mark after the name tag. We generally do that in releases, however, throughout the rest of this section we omit it. We should probably remove this to be consistent throughout the whole release version section.